mozilla/fx-private-relay
# MPP-3932: Disable masks on complaints, round 2

This replaces #5114, and goes a lot farther.

This rewrites `_handle_complaint`, moving most of the logic into three new helper functions:

* `_get_complaint_data` gets the relevant data out of the complaint notification, with lots of checks and logging if the message is not what we're expecting.
* `_gather_complainers` cross-checks the sender and receiver data against the database, getting the related users and masks (`RelayAddress`, `DomainAddress`). It also checks for the developer "simulate_complaint" mode, and swaps the real user email back in.
* `_reduce_future_complaints` takes the actions to reduce future complaints.
  - On the first complaint, turns on the `auto_block_spam` flag to drop emails that AWS SES identifies as spam (same as before)
  - On the second complaint, it turns the associated mask to block emails and sends an email to the user. This has been altered to look at the `From:` header data, which has the mask, rather than the `source` field, which is our reply address.
  - On future complaints (`how?`), it does nothing but takes note that it did nothing (new)

This leaves `_handle_complaint` to log the complaint in developer mode, log what happened for each detected Relay user, and emit a counter.

I also extended the developer mode to correctly work for domain masks, added `_get_address_if_exists`, removed a `mypy` override, and fixed Heroku Redis... This PR can definitely be broken into smaller bits. 

# How to test:

The test suite has been expanded to cover the new code. There are lots of new tests.

I pushed this branch to Heroku for manual testing. There are two issues ~I've seen~ that are now fixed on Heroku:

* Each forwarded email appears to result in two complaints - Misconfiguration of SES on dev server
* In the email "Firefox Relay⁩ has deactivated one of your email masks", there's some extra `???` and the links do not work. - Fixed the HTML template

Here's the test process:

* Deploy this branch
* Send a test email to a mask. It should be forwarded.
* Load the Papertrail app in Heroku. Filter on `program:"app/worker.1"`. You'll see logs related to your email.
* Add the `developer_mode` flag to your Heroku user. You can run a Heroku bash shell:
  `heroku run --app fx-private-relay bash`
  then run
  `./manage.py waffle_flag developer_mode --append --user "<your_real_email@example.com>"`
* Create a random mask, give it the description "DEV:simulate_complaint" (no spaces)
* Send an email to that mask
  - The email is _not_ delivered to you, but instead to the [AWS SES Complaint Simulator address](https://docs.aws.amazon.com/ses/latest/dg/send-an-email-from-console.html#send-email-simulator)
  - You should see several logs with `"msg": "_handle_received: developer_mode"`. The `"notification_gza85"` fields can be combined and passed to `emails.utils.decode_dict_gza85` to get the original delivery notification.
  - After a few more messages, you should see a log with `"msg": "_handle_complaint: developer mode"`. The `"notification_gza85"` fields can be combined and passed to `emails.utils.decode_dict_gza85` to get the original complaint notification.
  - The next log is `"msg": "complaint_notification"`, and has the standard log for a complaint. The `"relay_action"` should be `"auto_block_spam"`
  - There will be other logs as well, but these are the important ones.
* Send a second email to that mask
  - The logs repeat, with `"relay_action"` as `no_action` since `auto_block_spam` is already on.
* Add the `disable_mask_on_complaint` flag to your Heroku user. You can run a Heroku bash shell:
  `heroku run --app fx-private-relay bash`
  then run
  `./manage.py waffle_flag disable_mask_on_complaint --append --user "<your_real_email@example.com>"`
* Send a third email to that mask.
  - The logs repeat. The `"relay_action"` is `"disable_mask"` (and then `no_action` if there is a second complaint)
* Check your email after a minute
  - You should get the email "Firefox Relay⁩ has deactivated one of your email masks." to your real address. The links should go to the email dashboard.
* Go to the Email Dashboard, refreshing if needed (such as going to FAQ and back to the dashboard)
  - The mask is now set to Blocking
